### PR TITLE
Phase2-sim26 Add one simproducer class which can save hits of calorimeters

### PR DESCRIPTION
**Pull request by bsunanda**
It is the first step for comparing scoring in GeantV. This class using SimProducer mechanism can score hits with a geometry declaration having no Sensitive Detector. Being a SimProducer it does not interfere with standard SIM workflows

### CODE REVIEW

**Comment by cvuosalo** [comment]
+1

File: SimG4CMS/Calo/interface/CaloSteppingAction.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  double   curve_LY(const G4LogicalVolume* lv, double z);
+  double   getBirkL3(const G4Step* aStep);
+  double   getBirkHC(const G4Step* aStep);
+  void     saveHits(int flag);
+
+  std::unique_ptr<EcalBarrelNumberingScheme> ebNumberingScheme_;
+  std::unique_ptr<EcalEndcapNumberingScheme> eeNumberingScheme_;
+  std::unique_ptr<HcalNumberingFromDDD>      hcNumbering_;
+  std::unique_ptr<HcalNumberingScheme>       hcNumberingScheme_;
+  std::unique_ptr<CaloSlaveSD>               slave_[3];
```
Comment (kpedro88) [inline SimG4CMS/Calo/interface/CaloSteppingAction.h]: better to make magic number 3 a named constant and use it everywhere (also below and in .cc file)

File: SimG4CMS/Calo/src/CaloSteppingAction.cc [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+			   << ":" << birkSlopeEC_ << ":" << birkCutEC_;
+  edm::LogVerbatim("Step") << "CaloSteppingAction::Constants for HCAL: Birk "
+			   << "constants " << birkC1HC_ << ":" << birkC2HC_
+			   << ":" << birkC3HC_;
+  edm::LogVerbatim("Step") << "CaloSteppingAction:: " << nameHitC_.size() 
+			   << " hit collections";
+  for (unsigned int k=0; k<nameHitC_.size(); ++k)
+    edm::LogVerbatim("Step") << "[" << k << "] " << nameHitC_[k];
+
+  ebNumberingScheme_.reset(new EcalBarrelNumberingScheme());
```
Comment (kpedro88) [inline SimG4CMS/Calo/src/CaloSteppingAction.cc]: better to use `ebNumberingScheme_ = std::make_unique<EcalBarrelNumberingScheme>();` and similar for all `unique_ptr`s below

File: SimG4CMS/Calo/src/CaloSteppingAction.cc [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+  edm::LogVerbatim("Step") << "CaloSteppingAction: -------->  Total number of "
+			   << "selected entries : " << count_;
+}
+
+void CaloSteppingAction::produce(edm::Event& e, const edm::EventSetup&) {
+
+  for (int k=0; k<3; ++k) {
+    saveHits(k);
+    std::unique_ptr<edm::PCaloHitContainer>
+      product(new edm::PCaloHitContainer);
```
Comment (kpedro88) [inline SimG4CMS/Calo/src/CaloSteppingAction.cc]: `auto product = std::make_unique<edm::PCaloHitContainer>();`

File: SimG4CMS/Calo/src/CaloSteppingAction.cc [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+void CaloSteppingAction::update(const BeginOfJob * job) {
+  edm::LogVerbatim("Step") << "CaloSteppingAction:: Enter BeginOfJob";
+
+  // Numbering From DDD
+  edm::ESHandle<HcalDDDSimConstants>    hdc;
+  (*job)()->get<HcalSimNumberingRecord>().get(hdc);
+  const HcalDDDSimConstants* hcons_ = hdc.product();
+  edm::LogVerbatim("Step") << "CaloSteppingAction:: Initialise "
+			   << "HcalNumberingFromDDD";
+  hcNumbering_.reset(new HcalNumberingFromDDD(hcons_));
```
Comment (kpedro88) [inline SimG4CMS/Calo/src/CaloSteppingAction.cc]: use `make_unique`

Comment (bsunanda) [reply SimG4CMS/Calo/src/CaloSteppingAction.cc]: Done

File: SimG4CMS/Calo/src/CaloSteppingAction.cc (hunk 3 above)
Comment (bsunanda) [reply SimG4CMS/Calo/src/CaloSteppingAction.cc]: Done

File: SimG4CMS/Calo/src/CaloSteppingAction.cc (hunk 2 above)
Comment (bsunanda) [reply SimG4CMS/Calo/src/CaloSteppingAction.cc]: Done

File: SimG4CMS/Calo/src/CaloSteppingAction.cc [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+			   << " hit collections";
+  for (unsigned int k=0; k<nameHitC_.size(); ++k)
+    edm::LogVerbatim("Step") << "[" << k << "] " << nameHitC_[k];
+
+  ebNumberingScheme_ = std::make_unique<EcalBarrelNumberingScheme>();
+  eeNumberingScheme_ = std::make_unique<EcalEndcapNumberingScheme>();
+  hcNumbering_.reset(nullptr);
+  hcNumberingScheme_ = std::make_unique<HcalNumberingScheme>();
+  for (int k=0; k<3; ++k) {
+    slave_[k].reset(new CaloSlaveSD(nameHitC_[k]));
```
Comment (kpedro88) [inline SimG4CMS/Calo/src/CaloSteppingAction.cc]: also use `make_unique` here

File: SimG4CMS/Calo/interface/CaloSteppingAction.h (hunk 1 above)
Comment (bsunanda) [reply SimG4CMS/Calo/interface/CaloSteppingAction.h]: Done

File: SimG4CMS/Calo/src/CaloSteppingAction.cc (hunk 5 above)
Comment (bsunanda) [reply SimG4CMS/Calo/src/CaloSteppingAction.cc]: Done

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by cvuosalo** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**